locust/dispatch.py

```diff
@@ -49,548 +49,241 @@ class UsersDispatcher(Iterator):
 
-        workers_user_count = self._workers_user_count
+        self._wait_between_dispatch = self._user_count_per_dispatch_iteration / self._spawn_rate
 
-        # Represents the ideal workers on which we'd want to add the user class
-        # because these workers contain less users than all the other workers
-        ideal_worker_node_ids = [
-            ideal_worker_node_id
-            for ideal_worker_node_id in workers_user_count.keys()
-            if workers_user_count[ideal_worker_node_id] + 1 - min(workers_user_count.values()) < 2
-        ]
+        self._initial_users_on_workers = self._users_on_workers
 
-        if worker_node_id_to_add_user_on in ideal_worker_node_ids:
-            return False
+        self._users_on_workers = self._fast_users_on_workers_copy(self._initial_users_on_workers)
 
-        # Only keep the workers having less users than the target value as
-        # we can't add users to those workers anyway.
-        workers_user_count_without_excess_users = {
-            worker_node_id: user_count
-            for worker_node_id, user_count in workers_user_count.items()
-            if user_count < sum(self._desired_users_assigned_to_workers[worker_node_id].values())
-        }
+        self._current_user_count = sum(map(sum, map(dict.values, self._users_on_workers.values())))
 
-        ideal_worker_on_which_to_add_user_exists = any(
-            self._users_left_to_assigned[ideal_worker_node_id][user_class] > 0
-            for ideal_worker_node_id in ideal_worker_node_ids
-        )
+        self._dispatcher_generator = self._dispatcher()
 
-        if worker_node_id_to_add_user_on not in workers_user_count_without_excess_users:
-            return ideal_worker_on_which_to_add_user_exists
+        self._dispatch_iteration_durations.clear()
 
-        if (
-            workers_user_count_without_excess_users[worker_node_id_to_add_user_on]
-            + 1
-            - min(workers_user_count.values())
-            >= 2
-            and ideal_worker_on_which_to_add_user_exists
-        ):
-            # Adding the user to the current worker will result in this worker having more than 1
-            # extra users compared to the other workers (condition on the left of the `and` above).
-            # Moreover, we know there exists at least one other worker that would better host
-            # the new user (condition on the right of the `and` above). Thus, we skip to the next worker node.
-            return True
+    def add_worker(self, worker_node: "WorkerNode") -> None:
+        self._worker_nodes.append(worker_node)
+        self._worker_nodes = sorted(self._worker_nodes, key=lambda w: w.id)
+        self._prepare_rebalance()
 
-        return False
+    def remove_worker(self, worker_node_id: str) -> None:
+        self._worker_nodes = [w for w in self._worker_nodes if w.id != worker_node_id]
+        if len(self._worker_nodes) == 0:
+            # TODO: Test this
```

Comment:
There is a test for this now, right? Remove the todo :)

Reply:
I don't think there's a test for the zero worker case yet. I'll add one.